Fix Windows guards in c10d

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59563 Move c10d to libtorch(_cuda)
* #59697 Add guards for USE_C10D_FOO in relevant c10d files
* **#59696 Fix Windows guards in c10d**
* #59562 Remove USE_CUDA from c10d reducer/logger
* #59561 Remove USE_CUDA from ProcessGroupGloo

Some files in c10d refer to dist autograd. However, on Windows, dist autograd isn't built. Hence we need to "mask out" those references under Windows. This was already partly done, but when moving c10d to libtorch some issues came up, possibly due to the different way in which linking happens. Hence I masked out the remaining references.

Differential Revision: [D28987579](https://our.internmc.facebook.com/intern/diff/D28987579/)

lw committed:
Fix Windows guards in c10d

Some files in c10d refer to dist autograd. However, on Windows, dist autograd isn't built. Hence we need to "mask out" those references under Windows. This was already partly done, but when moving c10d to libtorch some issues came up, possibly due to the different way in which linking happens. Hence I masked out the remaining references.

Differential Revision: [D28987579](https://our.internmc.facebook.com/intern/diff/D28987579/)

[ghstack-poisoned]

lw committed:
Update on "Fix Windows guards in c10d"


Some files in c10d refer to dist autograd. However, on Windows, dist autograd isn't built. Hence we need to "mask out" those references under Windows. This was already partly done, but when moving c10d to libtorch some issues came up, possibly due to the different way in which linking happens. Hence I masked out the remaining references.

Differential Revision: [D28987579](https://our.internmc.facebook.com/intern/diff/D28987579/)

[ghstack-poisoned]

cbalioglu commented on torch/lib/c10d/reducer.hpp:
```diff
@@ -225,8 +229,12 @@ class TORCH_API Reducer {
-  using GradCallback =
-      torch::distributed::autograd::DistAutogradContext::GradCallback;
+  using GradCallback = std::function<bool(at::Tensor&)>;
+#ifndef _WIN32
```
Why do we need a `static_assert` specifically for Windows? And as far as I understand both template arguments to `std::is_same<>` are concrete types. Overall what is the purpose of this assertion?

lw replied:
We "need" a static assert for when we're _not_ on Windows. The reasoning is: we want this `GradCallback` here to be an alias for whatever dist autograd's `GradCallback` type is, however under Windows we don't have access to dist autograd. Hence we "resolve" this local alias here to its concrete type, but we double-check that it still matches the dist autograd type when dist autograd is available. (We do it through a `static_assert`, rather than through a comment, because `static_assert`s are self-enforcing and never go stale).

lw committed:
Update on "Fix Windows guards in c10d"


Some files in c10d refer to dist autograd. However, on Windows, dist autograd isn't built. Hence we need to "mask out" those references under Windows. This was already partly done, but when moving c10d to libtorch some issues came up, possibly due to the different way in which linking happens. Hence I masked out the remaining references.

Differential Revision: [D28987579](https://our.internmc.facebook.com/intern/diff/D28987579/)

[ghstack-poisoned]

lw committed:
Update on "Fix Windows guards in c10d"


Some files in c10d refer to dist autograd. However, on Windows, dist autograd isn't built. Hence we need to "mask out" those references under Windows. This was already partly done, but when moving c10d to libtorch some issues came up, possibly due to the different way in which linking happens. Hence I masked out the remaining references.

Differential Revision: [D28987579](https://our.internmc.facebook.com/intern/diff/D28987579/)

[ghstack-poisoned]

lw committed:
Update on "Fix Windows guards in c10d"


Some files in c10d refer to dist autograd. However, on Windows, dist autograd isn't built. Hence we need to "mask out" those references under Windows. This was already partly done, but when moving c10d to libtorch some issues came up, possibly due to the different way in which linking happens. Hence I masked out the remaining references.

Differential Revision: [D28987579](https://our.internmc.facebook.com/intern/diff/D28987579/)

[ghstack-poisoned]

agolynski commented on torch/lib/c10d/reducer.hpp:
```diff
@@ -224,8 +228,12 @@ class Reducer {
+#ifndef _WIN32
+  static_assert(std::is_same<
+    GradCallback,
+    torch::distributed::autograd::DistAutogradContext::GradCallback>::value, "");
```
Could you add a message here?

lw committed:
Update on "Fix Windows guards in c10d"


Some files in c10d refer to dist autograd. However, on Windows, dist autograd isn't built. Hence we need to "mask out" those references under Windows. This was already partly done, but when moving c10d to libtorch some issues came up, possibly due to the different way in which linking happens. Hence I masked out the remaining references.

Differential Revision: [D28987579](https://our.internmc.facebook.com/intern/diff/D28987579/)

[ghstack-poisoned]
